### PR TITLE
[1.x] Support Laravel 8, Abandon tenancy/testing

### PR DESCRIPTION
**Pull request by ArlonAntonius**

### CODE REVIEW

File: .github/workflows/code-style.yml [hunk 1]
```diff
@@ -1,21 +0,0 @@
...
-            - name: Select PHP version
-              uses: shivammathur/setup-php@master
-              with:
-                  php-version: '7.3'
-
-            - name: Install Composer dependencies
-              run: composer install
-
-            - name: Run tests for coverage
-              run: ./vendor/bin/phpcs
```
Comment (ArlonAntonius) [inline .github/workflows/code-style.yml:21]: Will add this back at some point :)

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -23,6 +24,10 @@ jobs:
                       db: MySQL5
                     - service: mariadb
                       db: MariaDB
+                
+                exclude:
+                  - laravel: 8.*
```
Comment (ArlonAntonius) [inline .github/workflows/test.yml:29]: L8 Requires PHP >= 7.3

File: .styleci.yml [hunk 3]
```diff
@@ -1,3 +1,4 @@
 preset: recommended
+risky: true
```
Comment (ArlonAntonius) [inline .styleci.yml:2]: Allows StyleCI to add the declare at the top of the file

File: build/tenancy-split.sh [hunk 4]
```diff
@@ -33,7 +33,7 @@ git subsplit publish --heads="master" src/Identification/Queue:git@github.com:te
 git subsplit publish --heads="master" src/Database/Mysql:git@github.com:tenancy/db-driver-mysql.git
 git subsplit publish --heads="master" src/Database/Sqlite:git@github.com:tenancy/db-driver-sqlite.git
 
-# Testing
-git subsplit publish --heads="master" src/Testing:git@github.com:tenancy/testing.git
+# Testing (Abandoned)
+# git subsplit publish --heads="master" src/Testing:git@github.com:tenancy/testing.git
```
Comment (ArlonAntonius) [inline build/tenancy-split.sh:37]: Will do this one manually once after release

File: composer.json [hunk 5]
```diff
@@ -4,17 +4,14 @@
...
-        "fzaninotto/faker": "^1.7",
-        "laravel/laravel": "^7.0",
+        "fzaninotto/faker": "^1.9",
+        "laravel/laravel": "^7.0 | ^8.0",
         "phpunit/phpunit": "^8.0",
-        "mockery/mockery": "^1.0",
-        "squizlabs/php_codesniffer": "^3.3",
-        "psalm/plugin-laravel": "^1.1",
-        "laminas/laminas-diactoros": "^1.0"
+        "mockery/mockery": "^1.0"
```
Comment (ArlonAntonius) [inline composer.json:14]: Removed tons of things we don't use anymore

File: src/Testing/Concerns/CreatesApplication.php [hunk 6]
```diff
@@ -83,12 +84,29 @@ protected function bootTenancy()
...
+            return;
+        }
+
+        \Illuminate\Database\Eloquent\Factories\Factory::guessFactoryNamesUsing(function (string $modelName) {
+            if (is_subclass_of((new $modelName()), Tenant::class) || $modelName === Tenant::class) {
+                return TenantFactory::class;
+            }
+
+            throw new \InvalidArgumentException('This is only meant to be used by Tenancy');
+        });
```
Comment (ArlonAntonius) [inline src/Testing/Concerns/CreatesApplication.php:109]: As we're only using these files internally now, I'm throwing a hard error :)

File: src/Testing/Mocks/Tenant.php [hunk 7]
```diff
@@ -31,4 +31,27 @@ class Tenant extends Model implements Contract
...
+    public static function factory(...$parameters)
+    {
+        if (function_exists('factory')) {
+            return factory(get_called_class(), $parameters);
+        }
+
+        return \Illuminate\Database\Eloquent\Factories\Factory::factoryForModel(get_called_class())
+                    ->count(is_numeric($parameters[0] ?? null) ? $parameters[0] : null)
+                    ->state(is_array($parameters[0] ?? null) ? $parameters[0] : ($parameters[1] ?? []));
+    }
```
Comment (ArlonAntonius) [inline src/Testing/Mocks/Tenant.php:44]: Basically:
- Support old factory usage
- Support new factory usage, by copy pasting the hasFactory trait :)

File: tests/UsesModels.php [hunk 8]
```diff
@@ -22,11 +22,19 @@ trait UsesModels
...
-        $factory->load(__DIR__.DIRECTORY_SEPARATOR.
-            'Mocks'.DIRECTORY_SEPARATOR.
-            'Models'.DIRECTORY_SEPARATOR.
-            'factories');
+        if (class_exists(\Illuminate\Database\Eloquent\Factory::class)) {
+            /** @var \Illuminate\Database\Eloquent\Factory $factory */
+            $factory = resolve(\Illuminate\Database\Eloquent\Factory::class);
+
+            /** @var Factory $factory */
+            $factory = resolve(Factory::class);
```
Comment (Plytas) [inline tests/UsesModels.php]: Aren't these 2 the same?

**Review by Plytas — CHANGES_REQUESTED** [review]

File: tests/UsesModels.php (hunk 8 above)
Comment (ArlonAntonius) [reply tests/UsesModels.php]: You are right, will change it asap

**Review by Plytas — DISMISSED** [review]

**Review by Plytas — DISMISSED** [review]

**Review by Plytas — APPROVED** [review]

**Review by fletch3555 — APPROVED** [review]

**Pull request merged**